src/Google.Api.Gax/Platform.cs

```
@@ -0,0 +1,150 @@
+        /// </summary>
+        /// <returns>A task containing the execution platform information.</returns>
+        public static Task<Platform> InstanceAsync() => s_instance.Value;
+
+        /// <summary>
+        /// Get execution platform information. This may block for up to one second.
+        /// </summary>
+        /// <returns>Execution platform information.</returns>
+        public static Platform Instance() => InstanceAsync().Result;
+
+        private static async Task<GcePlatformDetails> LoadGceDetails()
+        {
+            const string metadataUrl = "http://metadata.google.internal/computeMetadata/v1?recursive=true";
+            const string metadataFlavorKey = "Metadata-Flavor";
+            const string metadataFlavorValue = "Google";
+            const long maxContentLength = 512 * 1024; // Maximum allowed metadata size.
+            var httpRequest = new HttpRequestMessage(HttpMethod.Get, metadataUrl);
+            httpRequest.Headers.Add(metadataFlavorKey, metadataFlavorValue); // Required for any query.
+            var cts = new CancellationTokenSource(TimeSpan.FromMilliseconds(1000)); // 1000 found to be reasonable.
+            var httpClient = new HttpClient();
+            var response = await httpClient.SendAsync(httpRequest, cts.Token);
+            IEnumerable<string> metadataValues;
+            if (response.StatusCode == HttpStatusCode.OK
+                && response.Content.Headers.TryGetValues(metadataFlavorKey, out metadataValues)
+                && metadataValues.Contains(metadataFlavorValue)
+                && response.Content.Headers.ContentLength < maxContentLength)
+            {
+                // Valid response from metadata server.
+                string metadataJson = await response.Content.ReadAsStringAsync();
+                try
+                {
+                    return new GcePlatformDetails(metadataJson);
+                }
+                catch
```

In general, we do not catch all exceptions.